app/helpers/fhir_helper.rb

```diff
@@ -232,6 +232,67 @@ def close_contact_from_fhir(related_person)
     }
   end
 
+  # Returns a representative FHIR::Immunization for an instance of a Sara Alert Vaccine.
+  # https://www.hl7.org/fhir/immunization.html
+  def vaccine_as_fhir(vaccine)
+    FHIR::Immunization.new(
+      meta: FHIR::Meta.new(lastUpdated: vaccine.updated_at.strftime('%FT%T%:z')),
+      id: vaccine.id,
+      status: 'completed',
+      vaccineCode: [
+        FHIR::CodeableConcept.new(
+          text: vaccine.product_name,
+          coding: Vaccine.product_codes_by_name(vaccine.group_name, vaccine.product_name).map do |code|
+            FHIR::Coding.new(code: code['code'], system: code['system'])
+          end
+        )
+      ],
+      patient: FHIR::Reference.new(reference: "Patient/#{vaccine.patient_id}"),
+      occurrenceDateTime: vaccine.administration_date,
+      note: FHIR::Annotation.new(text: vaccine.notes),
+      protocolApplied: [
+        {
+          targetDisease: [
+            FHIR::CodeableConcept.new(
+              text: vaccine.group_name,
+              coding: Vaccine.group_codes_by_name(vaccine.group_name).map do |code|
+                FHIR::Coding.new(code: code['code'], system: code['system'])
+              end
+            )
+          ],
+          doseNumberString: vaccine.dose_number
+        }
+      ]
+    )
+  end
+
+  # Create a hash of atttributes that corresponds to a Sara Alert Patient (and can be used to
+  # create new ones, or update existing ones), using the given FHIR::Patient.
+  # Hash is of the form:
+  # {
+  #  attribute_name: { value: <converted-value>, path: <fhirpath-to-corresponding-fhir-element>, errors: <array-of-messages> }
+  # }
+  def vaccine_from_fhir(vaccine)
+    group_coding = vaccine&.protocolApplied&.first&.targetDisease&.first&.coding&.first
+    group = Vaccine.group_name_by_code(group_coding&.system, group_coding&.code)
+    group_errors = ["is not an acceptable value, acceptable values are: #{Vaccine.group_code_options.join(', ')}"] if group.nil? && !group_coding.nil?
+
+    product_name_coding = vaccine&.vaccineCode&.first&.coding&.first
+    product_name = Vaccine.product_name_by_code(group, product_name_coding&.system, product_name_coding&.code)
+    if product_name.nil? && !product_name_coding.nil?
+      product_name_errors = ["is not an acceptable value, acceptable values are: #{Vaccine.product_code_options(group).join(', ')}"]
```

Comment:
I'm thinking we should skip the product error if there's a group error. If I send an invalid group, I get:
```
"Value '{\"system\"=>\"http://hl7.org/fhir/sid/cvx\", \"code\"=>\"207\"}' for 'Product Name' is not an acceptable value, acceptable values are:"
```
in addition to the group error, which seems confusing.

Reply:
This is addressed in https://github.com/SaraAlert/SaraAlert/commit/9ed52ca5e51fd889375b1caee4fa5a844f7c2853. Note that I also had to update the validator on the vaccine model for `product_name`.

Reply:
Thank you